src/hipscat_import/catalog/resume_plan.py

```diff
@@ -0,0 +1,224 @@
+
+            ## Validate existing resume state.
+            ## - if a later stage is complete, the earlier stages should be complete too.
+            mapping_done = self.is_mapping_done()
+            splitting_done = self.is_splitting_done()
+            reducing_done = self.is_reducing_done()
+
+            if reducing_done and (not mapping_done or not splitting_done):
+                raise ValueError(
+                    "mapping and splitting must be complete before reducing"
+                )
+            if splitting_done and not mapping_done:
+                raise ValueError("mapping must be complete before splitting")
+            step_progress.update(1)
+
+            ## Gather keys for execution.
+            self.input_paths.sort()
+            for test_path in self.input_paths:
+                if not file_io.does_file_or_directory_exist(test_path):
+                    raise FileNotFoundError(f"{test_path} not found on local storage")
+            step_progress.update(1)
+            if not mapping_done:
+                mapping_start_keys = self._read_log_keys(self.MAPPING_START_LOG_FILE)
+                mapping_done_keys = self._read_log_keys(self.MAPPING_DONE_LOG_FILE)
+                if len(mapping_done_keys) != len(mapping_start_keys):
+                    raise ValueError(
+                        "Resume logs are corrupted. "
+                        "Delete temp directory and restart import pipeline."
+                    )
+                mapped_paths = set(mapping_start_keys)
+                self.map_files = [
+                    file_path
+                    for file_path in self.input_paths
+                    if f"map_{file_path}" not in mapped_paths
+                ]
+            if not splitting_done:
+                split_keys = set(self._read_log_keys(self.SPLITTING_LOG_FILE))
+                self.split_keys = [
+                    (f"split_{i}", file_path)
+                    for i, file_path in enumerate(self.input_paths)
+                ]
+                self.split_keys = [
+                    (key, file)
+                    for (key, file) in self.split_keys
+                    if key not in split_keys
+                ]
+            ## We don't pre-gather the plan for the reducing keys.
+            ## It requires the full destination pixel map.
+            step_progress.update(1)
+
+    def read_histogram(self, highest_healpix_order):
+        """Read a numpy array at the indicated directory.
+        Otherwise, return histogram of appropriate shape."""
+        file_name = file_io.append_paths_to_pointer(
+            self.tmp_path, self.HISTOGRAM_BINARY_FILE
+        )
+        if file_io.does_file_or_directory_exist(file_name):
+            with open(file_name, "rb") as file_handle:
+                return frombuffer(file_handle.read(), dtype=np.int64)
+        return pixel_math.empty_histogram(highest_healpix_order)
+
+    def mark_mapping_done(self, mapping_key: str, histogram):
```

Comment:
I feel like this name is a little confusing given it's also writing the histogram as well as marking it done